normandy/selfrepair/static/js/self_repair_driver.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * Convenience wrapper around XMLHttpRequest.
+ *
+ * @param {String} method - HTTP method to use.
+ * @param {String} url - URL to send the request to.
+ * @param {Object} options - Optional arguments.
+ * @param {Object} options.data - Request data as an Object; keys are
+ *     used as parameter names, values as parameter values.
+ * @param {Object} options.headers - Headers and values to set on the
+ *     request.
+ * @promise {XMLHttpRequest} The request after it has succeeded.
+ * @rejects {XMLHttpRequest} The request after it has failed.
+ */
+function xhr(method, url, options={}) {
+    return new Promise((resolve, reject) => {
+        let req = new XMLHttpRequest();
+        req.addEventListener('load', e => {
+            if (req.status !== 200) {
+                reject(req);
+            } else {
+                resolve(req);
+            }
+        });
+        req.open(method, url);
+
+        let data = undefined;
+        if (options.data) {
+            req.setRequestHeader('Content-Type', 'application/json');
+            data = JSON.stringify(data);
+        }
+
+        if (options.headers) {
+            for (let key in options.headers) {
+                req.setRequestHeader(key, options.headers[key]);
+            }
+        }
+
+        req.send(data);
+    });
+}
+xhr.get = xhr.bind(null, 'GET');
+xhr.post = xhr.bind(null, 'POST');
+
+
+let registeredActions = {};
+window.Normandy = {
+    registerAction(name, func) {
+        registeredActions[name] = func;
+    }
+};
+
+
+/**
+ * Download the implementation of the given action from the server.
+ *
+ * @param {Action} action  - Action taken from a Recipe.
+ * @promise Resolves once the action has been loaded and registered.
+ * @rejects {Error} Rejects if the action could not be loaded or did not
+ *     register itself.
+ */
+function loadAction(action) {
+    return new Promise((resolve, reject) => {
+        if (!registeredActions[action.name]) {
+            let script = document.createElement('script');
+            script.src = action.implementation.url;
+            script.onload = () => {
+                if (!registeredActions[action.name]) {
+                    reject(new Error(`Could not find action with name ${action.name}.`));
+                } else {
+                    resolve();
```

Comment:
I think this function should resolve with the action that was loaded. That way other things, like `runRecipe` can be simplified a little to not have to deal with the global registry.